### PR TITLE
MongoDB benchmark

### PR DESCRIPTION
**Pull request by anish-palakurthi**
Support for running the MongoDB benchmark within KernMLOps. See `README.md` for execution instructions.

### CODE REVIEW

File: python/kernmlops/kernmlops_benchmark/mongodb.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  BenchmarkNotRunningError,
+  BenchmarkRunningError,
+)
+from kernmlops_config import ConfigBase
+
+
+@dataclass(frozen=True)
+class BenchmarkConfig(ConfigBase):
+  
+  recordCount: int = 1000000
```
Comment (patrickkenney9801) [inline python/kernmlops/kernmlops_benchmark/mongodb.py:18]: please use lower snake case instead of camel case, i.e. `record_count`
note: Pascal case is fine for class names as you have done

File: python/kernmlops/kernmlops_benchmark/mongodb.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        return "mongodb"
+
+    @classmethod
+    def default_config(cls) -> ConfigBase:
+        return BenchmarkConfig()
+
+    @classmethod
+    def from_config(cls, config: ConfigBase) -> "Benchmark":
+        generic_config = cast(GenericBenchmarkConfig, getattr(config, "generic"))
+        gap_config = cast(BenchmarkConfig, getattr(config, cls.name()))
```
Comment (patrickkenney9801) [inline python/kernmlops/kernmlops_benchmark/mongodb.py:36]: please rename this variable to `mongodb_config`

File: python/kernmlops/kernmlops_benchmark/mongodb.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    def run(self) -> None:
+        if self.process is not None:
+            raise BenchmarkRunningError()
+
+        bash_file_path = self.benchmark_dir / "run_mongodb.sh" # Add the path to your bash file here
+        print(bash_file_path)
+        self.process = subprocess.Popen(
+            [
+                "bash",
+                bash_file_path,
```
Comment (patrickkenney9801) [inline python/kernmlops/kernmlops_benchmark/mongodb.py:64]: please cast this `Path` type to a string before passing it, i.e. `str(bash_file_path)`

File: run_gapbs.sh [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    log "Benchmark suite completed successfully!"
+    log "Results are available in $LOG_FILE"
+    log "Detailed report available in report_$(date +%Y%m%d_%H%M%S).txt"
+}
+
+# Trap Ctrl+C and call cleanup
+trap cleanup INT
+
+# Run main function
+main
```
Comment (patrickkenney9801) [inline run_gapbs.sh:182]: it looks like you have not installed any `pre-commit` hooks,
please run `make hooks` and then `make pre-commit`, fixing any issues that appear -- although that command will fix some of the simpler ones like adding a newline at the end of every file, in those cases simply do `git add . && make pre-commit`

File: run_redis.sh [hunk 5]
```diff
@@ -0,0 +1,370 @@
+#!/bin/bash
+
```
Comment (patrickkenney9801) [inline run_redis.sh:2]: also, none of these scripts, if they are kept, should be in the top-level directory of this repo, please move them to a sane place under scripts for now, like `scripts/run-benchmarks/*`

**Comment by patrickkenney9801** [comment]
Please also run `make format` followed by `make lint` and fix any issues that appear in the python code

**Comment by patrickkenney9801** [comment]
We should probably meet sometime early next week to discuss what all of these scripts do and why they are needed in the codebase

File: python/kernmlops/kernmlops_benchmark/mongodb.py (hunk 1 above)
Comment (patrickkenney9801) [reply python/kernmlops/kernmlops_benchmark/mongodb.py:18]: Also, after you make the above fix, please run `make defaults` to generate a new `defaults.yaml` and then commit those changes

Comment (patrickkenney9801) [reply python/kernmlops/kernmlops_benchmark/mongodb.py:18]: Also, it does not look like you use these configuration options at all anywhere
Please use them to configure the mongodb benchmark

File: run_gapbs.sh [hunk 6]
```diff
@@ -0,0 +1,182 @@
+#!/bin/bash
+
+# GAPBS Benchmark Runner
```
[the remark this answers is not in the document]
Comment (anish-palakurthi) [reply run_gapbs.sh:3]: Wrote it myself

File: python/kernmlops/kernmlops_benchmark/mongodb.py (hunk 1 above)
Comment (anish-palakurthi) [reply python/kernmlops/kernmlops_benchmark/mongodb.py]: make defaults is consistently failing with this:
```
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/anish/KernMLOps/python/kernmlops/__main__.py", line 3, in <module>
    import cli
  File "/home/anish/KernMLOps/python/kernmlops/cli/__init__.py", line 7, in <module>
    import data_collection
  File "/home/anish/KernMLOps/python/kernmlops/data_collection/__init__.py", line 4, in <module>
    from data_collection import bpf_instrumentation as bpf
  File "/home/anish/KernMLOps/python/kernmlops/data_collection/bpf_instrumentation/__init__.py", line 5, in <module>
    from data_collection.bpf_instrumentation.bpf_hook import BPFProgram
  File "/home/anish/KernMLOps/python/kernmlops/data_collection/bpf_instrumentation/bpf_hook.py", line 4, in <module>
    from data_schema import CollectionTable
  File "/home/anish/KernMLOps/python/kernmlops/data_schema/__init__.py", line 7, in <module>
    from data_schema import perf
  File "/home/anish/KernMLOps/python/kernmlops/data_schema/perf/__init__.py", line 3, in <module>
    from data_schema.perf.perf_schema import (
  File "/home/anish/KernMLOps/python/kernmlops/data_schema/perf/perf_schema.py", line 3, in <module>
    from typing import Protocol, override
ImportError: cannot import name 'override' from 'typing' (/usr/lib/python3.10/typing.py)

```

Comment (patrickkenney9801) [reply python/kernmlops/kernmlops_benchmark/mongodb.py]: please run this inside docker, i.e.
```
make docker
make defaults
```

File: python/kernmlops/kernmlops_benchmark/gap.py [hunk 7]
```diff
@@ -51,6 +51,8 @@ def setup(self) -> None:
     def run(self) -> None:
         if self.process is not None:
             raise BenchmarkRunningError()
+
+        print(str(self.benchmark_dir / self.config.gap_benchmark))
```
Comment (patrickkenney9801) [inline python/kernmlops/kernmlops_benchmark/gap.py:55]: please remove this stray line

File: python/kernmlops/kernmlops_benchmark/mongodb.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+  BenchmarkNotRunningError,
+  BenchmarkRunningError,
+)
+from kernmlops_config import ConfigBase
+
+
+@dataclass(frozen=True)
+class MongoDbConfig(ConfigBase):
+  record_count: int = 1000000
+  readProportion: float = 0.25
```
Comment (patrickkenney9801) [inline python/kernmlops/kernmlops_benchmark/mongodb.py:18]: please also fix these variable names

File: python/kernmlops/kernmlops_benchmark/mongodb.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    def kill(self) -> None:
+        if self.process is None:
+            raise BenchmarkNotRunningError()
+        self.process.terminate()
+
+    @classmethod
+    def plot_events(cls, graph_engine: GraphEngine) -> None:
+        if graph_engine.collection_data.benchmark != cls.name():
+            raise BenchmarkNotInCollectionData()
+        # TODO(Patrick): plot when a trial starts/ends
```
Comment (patrickkenney9801) [inline python/kernmlops/kernmlops_benchmark/mongodb.py:90]: please remove this stray comment

File: Dockerfile.dev [hunk 10]
```diff
@@ -50,6 +50,21 @@ RUN apt update -y \
...
+    echo "deb [ arch=amd64,arm64 signed-by=/usr/share/keyrings/mongodb-server-6.0.gpg ] https://repo.mongodb.org/apt/ubuntu jammy/mongodb-org/6.0 multiverse" | tee /etc/apt/sources.list.d/mongodb-org-6.0.list && \
+    apt-get update && \
+    apt-get install -y mongodb-org && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/*
+
+# Install OpenJDK 11
+RUN apt-get update && \
+    apt-get install -y openjdk-11-jdk && \
+    echo 'export PATH=$PATH:/usr/lib/jvm/java-11-openjdk-amd64/bin' >> /root/.bashrc
```
Comment (patrickkenney9801) [inline Dockerfile.dev:66]: please change this to `/root/.profile` to be more consistent with the dockerfile
also please move it down to where the other lines append to this file

File: python/kernmlops/kernmlops_benchmark/mongodb.py (hunk 8 above)
Comment (patrickkenney9801) [reply python/kernmlops/kernmlops_benchmark/mongodb.py:18]: please change these names to `read_proportion` and `update_proportion` then re-generate the `defaults.yaml` with `make defaults`

File: python/kernmlops/kernmlops_benchmark/mongodb.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    def run(self) -> None:
+        if self.process is not None:
+            raise BenchmarkRunningError()
+
+        bash_file_path = "scripts/run_benchmarks/run_mongodb.sh"
+        self.process = subprocess.Popen(
+            [
+                "bash",
+                str(bash_file_path),
+                str(self.config.record_count),
```
Comment (patrickkenney9801) [inline python/kernmlops/kernmlops_benchmark/mongodb.py:63]: looking at `scripts/run_benchmarks/run_mongodb.sh` it looks like these arguments are not used at all, please delete the bash file not that it is 1 line and recreate it here

File: scripts/setup-benchmarks/install_ycsb.sh [hunk 12]
```diff
@@ -0,0 +1,13 @@
+# Update and install required packages
+apt update
```
Comment (patrickkenney9801) [inline scripts/setup-benchmarks/install_ycsb.sh:2]: why is this here?

File: scripts/setup-benchmarks/install_ycsb.sh [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+
+# Download YCSB
+curl -O --location https://github.com/brianfrankcooper/YCSB/releases/download/0.17.0/ycsb-0.17.0.tar.gz
+tar xfvz ycsb-0.17.0.tar.gz
+cd ycsb-0.17.0
+
+# Copy contents of ycsb_runner.py to bin/ycsb
+sudo cp ../scripts/setup-benchmarks/ycsb_runner.py bin/ycsb
```
Comment (patrickkenney9801) [inline scripts/setup-benchmarks/install_ycsb.sh:10]: why does this need sudo?

File: scripts/setup-benchmarks/install_ycsb.sh [hunk 14]
```diff
@@ -0,0 +1,13 @@
+# Update and install required packages
+apt update
+
+# Download YCSB
+curl -O --location https://github.com/brianfrankcooper/YCSB/releases/download/0.17.0/ycsb-0.17.0.tar.gz
```
Comment (patrickkenney9801) [inline scripts/setup-benchmarks/install_ycsb.sh:5]: please install `ycsb` to the same location that the other benchmarks do like in https://github.com/utcs-scea/KernMLOps/blob/0614a75224479808e21321eac337c3ad9151f578/scripts/setup-benchmarks/setup-gap.sh#L11

File: scripts/setup-benchmarks/install_mongodb.sh [hunk 15]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+mkdir -p /data/db
+
+# Start MongoDB server without sysctl
+mongod --dbpath /data/db --fork --logpath /var/log/mongodb.log
```
Comment (patrickkenney9801) [inline scripts/setup-benchmarks/install_mongodb.sh:5]: this a great way to pre-load the data, but how it is setup putting the db into `/data/db` it looks as if it should be run inside a container but the README.md states it should be run outside
ideally this would be `$BENCHMARK_DIR/mongo_db` as in https://github.com/utcs-scea/KernMLOps/blob/0614a75224479808e21321eac337c3ad9151f578/scripts/setup-benchmarks/setup-gap.sh#L11

Comment (patrickkenney9801) [reply scripts/setup-benchmarks/install_mongodb.sh:5]: Then it could be run inside or outside the container and work well, I would also recommend calling this load script from the setup stage of the benchmark but check if the dir exists first, if so skip the load step

File: scripts/setup-benchmarks/install_ycsb.sh (hunk 13 above)
Comment (anish-palakurthi) [reply scripts/setup-benchmarks/install_ycsb.sh:10]: bin/ycsb needs elevated privileges to modify

File: scripts/setup-benchmarks/install_ycsb.sh (hunk 12 above)
Comment (anish-palakurthi) [reply scripts/setup-benchmarks/install_ycsb.sh:2]: removed

File: scripts/setup-benchmarks/install_ycsb.sh (hunk 13 above)
Comment (patrickkenney9801) [reply scripts/setup-benchmarks/install_ycsb.sh:10]: that is not true, I just confirmed by downloading and deleting that file

I suspect that you originally downloaded the tar file or unzipped it as root (especially since you had `apt update` that requires root), and in that case you would need root to modify the file

please try running this all without root and things should just work

File: scripts/setup-benchmarks/install_mongodb.sh (hunk 15 above)
Comment (anish-palakurthi) [reply scripts/setup-benchmarks/install_mongodb.sh:5]: So just modify it to check for existence before creating it.

File: python/kernmlops/kernmlops_benchmark/mongodb.py (hunk 8 above)
Comment (patrickkenney9801) [reply python/kernmlops/kernmlops_benchmark/mongodb.py:18]: this is not resolved, please change these names to read_proportion and update_proportion then re-generate the defaults.yaml with make defaults

File: python/kernmlops/kernmlops_benchmark/mongodb.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+        return MongoDbBenchmark(generic_config=generic_config, config=mongodb_config)
+
+    def __init__(self, *, generic_config: GenericBenchmarkConfig, config: MongoDbConfig):
+        self.generic_config = generic_config
+        self.config = config
+        self.benchmark_dir = self.generic_config.get_benchmark_dir() / "ycsb"
+        self.process: subprocess.Popen | None = None
+
+    def is_configured(self) -> bool:
+        return True
```
Comment (patrickkenney9801) [inline python/kernmlops/kernmlops_benchmark/mongodb.py:45]: please remove this and the comment so that the check goes through

File: python/kernmlops/kernmlops_benchmark/mongodb.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+        self.process = subprocess.Popen(
+            [
+                f"{self.benchmark_dir}/ycsb-0.17.0/bin/ycsb",
+                "run",
+                "mongodb",
+                "-s",
+                "-P",
+                f"{self.benchmark_dir}/ycsb-0.17.0/workloads/workloada",
+                "-p",
+                f"operationcount={self.config.record_count}",
```
Comment (patrickkenney9801) [inline python/kernmlops/kernmlops_benchmark/mongodb.py:67]: please change the config argument from `record_count` to `operation_count` to reflect how it is used

File: scripts/setup-benchmarks/install_mongodb.sh (hunk 15 above)
Comment (patrickkenney9801) [reply scripts/setup-benchmarks/install_mongodb.sh:15]: Please change the `--dbpath` to be `--dbpath "$YCSB_BENCHMARK_DIR/mongo_db` and update the `mkdir -p` command above to create this new directory

**Comment by anish-palakurthi** [comment]
If you want to run different mongodb workloads (besides workloada), this can be modified. Not critical right now and a small change.

File: Makefile [hunk 18]
```diff
@@ -106,6 +107,19 @@ benchmark-gap:
...
+		-c ${KERNMLOPS_CONFIG_FILE} \
+		--benchmark mongodb
+
+load-mongodb:
+	@echo "Loading MongoDB benchmark"
+	@python $(YCSB_BENCHMARK_DIR)/ycsb-0.17.0/bin/ycsb load mongodb -s \
+		-P "$(YCSB_BENCHMARK_DIR)/ycsb-0.17.0/workloads/workloada" \
+		-p recordcount=1000000 \
+		-p mongodb.url=mongodb://localhost:27017/ycsb
+
```
Comment (patrickkenney9801) [inline Makefile:122]: please add a new-directive here:

```
start-mongodb:
  @mkdir -p "$(YCSB_BENCHMARK_DIR)/mongo_db"
  @mongod --dbpath "$(YCSB_BENCHMARK_DIR)/mongo_db" --fork --logpath /var/log/mongodb.log || { echo "Error is expected, just means that the server was already running"; true; }
```

and then call it instead inside `benchmark-mongodb` and `load-mongodb`
via `@${MAKE} start-mongodb`

this will make the process more reliable

**Review by patrickkenney9801 — APPROVED** [review]
Terrific, thank you very much!
